lnwallet/channel.go

```diff
@@ -5340,6 +5340,39 @@ func (lc *LightningChannel) NumLocalUpdatesPendingOnRemote() uint64 {
 	return lc.localUpdateLog.logIndex - lastRemoteCommit.ourMessageIndex
 }
 
+// NumLocalUpdatesPendingOnLocal returns the number of local updates that still
+// need to be applied to the local commitment tx.
+func (lc *LightningChannel) NumLocalUpdatesPendingOnLocal() uint64 {
```

Comment:
Isn't it enough to simply know if we owe a commitment or if they do? `OweCommitment` and `NeedCommitment` seem to cover those cases. Still getting through the diff however.

Reply:
Unfortunately, no. See [this](https://github.com/lightning/bolts/pull/869#issuecomment-1851585049). There is still some instability in spec discussion and so the details of this PR aren't ultra high priority, but in the interest of not wasting time I've opted to implement it to the best of my understanding knowing full well that we may wish to throw some of this away.

I would *prefer* that this code change wasn't needed, however, as written, the best interpretation I have of the spec requires this.

Reply:
Confirmed here that NeedCommitment isn't sufficient as either of those could return true, even if there are no local updates in either case.

Reply:
Been re-loading this context into my brain, was also tempted to think we could  `XOR(Owe, Need)` to decide whether we can send `STFU` but I agree that it doesn't work out. 

Totally possible there's a smarter way to use them, but I went through the options [here](https://docs.google.com/spreadsheets/d/1oLjDKGZp_s8AYqLToTRGRp9iJHTPIXv50MuIDBTe-Rk/edit?usp=sharing) and couldn't come up with anything myself.

Reply:
Here's a proof for why it's impossible to use the facilities of `Need`/`Owe` as is:

1. The commitment facility has two (orthogonal) dimensions: whose update, and whose commit.
2. Need/Owe collapse the "whose update" dimension via a union
3. STFU requirements are trying to collapse along the *other* dimension: whose commit
4. When you encounter problems with orthogonal dimensions, you cannot construct one basis vector from any linear combination of the other basis vectors.